graphix/open_graph.py

```diff
@@ -0,0 +1,383 @@
+        >>> Measurement(0.1, "XY") == Measurement(0.0, "XY")
+        False
+        """
+        if not isinstance(other, Measurement):
+            return NotImplemented
+
+        return np.allclose(self.angle, other.angle) and self.plane == other.plane
+
+    def is_z_measurement(self) -> bool:
+        """Indicates whether it is a Z measurement
+
+        Example
+        -------
+        >>> from graphix.open_graph import Measurement
+        >>> Measurement(0.0, "XY").is_z_measurement()
+        True
+        >>> Measurement(0.0, "YZ").is_z_measurement()
+        False
+        >>> Measurement(0.1, "XY").is_z_measurement()
+        False
+        """
+        return np.allclose(self.angle, 0.0) and self.plane == "XY"
+
+
+class OpenGraph:
+    """Open graph contains the graph, measurement, and input and output
+    nodes. This is the graph we wish to implement deterministically
+
+    :param inside: the underlying graph state
+    :param measurements: a dictionary whose key is the ID of a node and the
+        value is the measurement at that node
+    :param inputs: a set of IDs of the nodes that are inputs to the graph
+    :param outputs: a set of IDs of the nodes that are outputs of the graph
+
+    Example
+    -------
+    >>> import networkx as nx
+    >>> from graphix.open_graph import OpenGraph, Measurement
+    >>>
+    >>> inside_graph = nx.Graph([(0, 1), (1, 2), (2, 0)])
+    >>>
+    >>> measurements = {i: Measurement(0.5 * i, "XY") for i in range(2)}
+    >>> inputs = [0]
+    >>> outputs = [2]
+    >>> og = OpenGraph(inside_graph, measurements, inputs, outputs)
+    """
+
+    # All the measurement, inputs and outputs information is stored in this
+    # graph. This is to simplify certain operations such as composition of
+    # graphs and checking equality.
+    # These attributes shouldn't be accessed directly by the user, instead,
+    # they should use the inputs(), outputs(), and measurements() properties.
+    #
+    # The following attributes are stored in the nodes:
+    #
+    # Non-output nodes:
+    # * "measurement" - The Measurement object for the node
+    #
+    # Input nodes:
+    # * "is_input"    - The value is always True
+    # * "input_order" - A zero-indexed integer used to preserve the ordering of
+    #                   the inputs
+    #
+    # Output nodes:
+    # * "is_output"    - The value is always True
+    # * "output_order" - A zero-indexed integer used to preserve the ordering of
+    #                    the outputs
+    _inside: nx.Graph
+
+    def __eq__(self, other):
+        """Checks the two open graphs are equal
+
+        This doesn't check they are equal up to an isomorphism"""
+
+        g1 = self.perform_z_deletions()
+        g2 = other.perform_z_deletions()
+
+        return (
+            g1.inputs == g2.inputs
+            and g1.outputs == g2.outputs
+            and nx.utils.graphs_equal(g1._inside, g2._inside)
+            and g1.measurements == g2.measurements
+        )
```

Comment:
I wonder what's the purpose of `perform_z_deletions` here. Does this become important in the context of FBQC you're working on (eg you end up having many Z measurements in the graph)? I get that this an easiest graph simplificaiton we can quickly implement, but it feels a little strange to include only this and not the other transformations like local complementation. If `perform_z_deletions` is necessary, we should mention in the docstring that `.. are equal up to the addition of Z-measured nodes`, and/or have separate method which explicitly does 'up to z measured nodes' comparison?

Reply:
Yes, it is more to do with some of the things we are doing in the compiler.
In this case since we aren't performing any optimisations, we don't need the `perform_z_deletions` methods here so I have removed them.